### PR TITLE
Accept { 'symbol literal': val } syntax

### LINKED ISSUE #1202: Symbol from String as Hash key
```ruby
hash = { 'test': :test }
```

throws exception

```
parse error on value ":" (tCOLON) :application:24
Source: application:24:14
sh = { 'test': :test }
~~~~~~~~~~~~~^
```

### PR DESCRIPTION
**Pull request by alexdowad**
This is intended to fix #1202.

BUT! But, test coverage is still needed! You can see I tried to add a test, but it seems that it's not even running. Can you point me to the best place to add tests for this syntax?

Rather than just testing that the parser returns the right AST, I'd like to actually test the semantics of expressions which use this syntax. I'd also like to add a test ensuring that the semantics of `cond ? a : b` are not broken.

### CODE REVIEW

**Comment by alexdowad** [comment]
Ooh, I forgot to regenerate the parser too. But tests come first.

**Comment by alexdowad** [comment]
I just slipped a fix for #462 in here as well. But likewise, it still needs tests!

**Comment by alexdowad** [comment]
Tried adding a test, but I am still not sure at all that these tests are even running. Some help, please!

**Comment by alexdowad** [comment]
BTW, I love the way you use a bit vector as a stack of `true` and `false` values. Bitshift to push a `false`, bitshift and bitwise AND to push a `true`, another bitshift to pop. `parser` could gain some performance by doing the same thing.

**Comment by elia** [comment]
The specs look good, a bit in a hurry now (7.50am) but you can look at CONTRIBUTING.md to see how to deal with and add rubyspec specs. 

All the cool stuff in the gramma/parser/lexer/compiler is almost entirely @adambeynon's work (props to him, once again)

**Comment by vais** [comment]
@alexdowad in order to run [the RubySpec's symbol spec](https://github.com/ruby/rubyspec/blob/master/language/symbol_spec.rb), just remove the following line from our `specs/rubyspecs` file: https://github.com/opal/opal/blob/master/spec/rubyspecs#L129

The `!` at the beginning of that line means that this specific file is currently *excluded* from spec runs even though the entire `rubyspec/language` directory is *included*. I tried explaining how this works [here](https://github.com/opal/opal/blob/master/CONTRIBUTING.md#down-the-rabbit-hole) :wink:

File: spec/lib/parser/unary_spec.rb [hunk 1]
```diff
@@ -5,6 +5,10 @@
     context 'with an integer' do
       it "parses unary op. with the right precedence" do
         parsed("-1.hello").should == [:call, [:int, -1], :hello, [:arglist]]
+        parsed("-1 ** 2").should == [:call,
+                                      [:call, [:int, 1], :**, [:arglist, [:int, 2]]],
+                                     :-@,
+                                     [:arglist]]
```
Comment (elia) [inline spec/lib/parser/unary_spec.rb]: It's probably better to keep this in a single line (not super important tho)

**Comment by alexdowad** [comment]
@vais, thanks! But that spec also contains the new `%i{}` syntax.

File: spec/lib/parser/unary_spec.rb (hunk 1 above)
Comment (alexdowad) [inline spec/lib/parser/unary_spec.rb]: Updated.

**Comment by alexdowad** [comment]
I found what looks like a good test at `spec/opal/core/language/versions/hash_1.9.rb`, but RSpec doesn't run it because it isn't suffixed with `_spec.rb`. Is this intentional? I have given it a `_spec.rb` suffix here, just to make it run, but I'm sure there is a better way to do this.

**Comment by elia** [comment]
That file should either match the spec pattern or be required by another file. 

If it adds good specs I'm ok with renaming it.

**Comment by elia** [comment]
Btw I thought the %i syntax was supported …

**Comment by alexdowad** [comment]
> I thought the %i syntax was supported …

It's implemented incorrectly:

```
20:30 ~prog/Ruby/opal % pry                                  [str_sym_literals_in_hash]
[1] pry(main)> $:.unshift Dir.pwd+'/lib'
=> ["/home/alex/Programming/Ruby/opal/lib",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/coderay-1.1.0/lib",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/slop-3.6.0/lib",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/method_source-0.8.2/lib",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/pry-0.10.3/lib",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/yard-0.8.7.6/lib",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/pry-doc-0.8.0/lib",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/site_ruby/2.2.0",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/site_ruby/2.2.0/x86_64-linux",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/site_ruby",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/vendor_ruby/2.2.0",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/vendor_ruby/2.2.0/x86_64-linux",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/vendor_ruby",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/2.2.0",
 "/home/alex/.rbenv/versions/2.2.2/lib/ruby/2.2.0/x86_64-linux"]
[2] pry(main)> require 'opal'
=> true
[3] pry(main)> Opal::Parser.new.parse("%i{a b c}")
=> (:array, (:str, "a"), (:str, "b"), (:str, "c"))
[4] pry(main)> Opal::Parser.new.parse("[:a, :b, :c]")                                   
=> (:array, (:sym, :a), (:sym, :b), (:sym, :c))
[5] pry(main)> %i{a b c}
=> [:a, :b, :c]
```

**Comment by alexdowad** [comment]
> That file should either match the spec pattern or be required by another file

Grepping for the original name of the file doesn't find anything...

And yes, it does add some good specs.

**Comment by elia** [comment]
%i has been probably implemented by extending the %q case, and since in opal strings and symbols are the same it shouldn't be a problem, but of course the implementation is not strictly "correct"

**Comment by alexdowad** [comment]
Thank you! If any further problems come up with this, please let me know so I can work on it more!

**Comment by elia** [comment]
Thank you for you help, you make changing the grammar look easy! 😸

**Pull request merged**